### PR TITLE
Dividing Iteration Classes

### PR DESCRIPTION
**Pull request by talbring**
## Proposed Changes

This divides the iteration files and adds a factory class for their creation. A rework of the iteration class is still necessary, but I don't have time to do that at the moment.

## Related Work
*Resolve any issues (bug fix or feature request), note any related PRs, or mention interactions with the work of others, if any.*

## PR Checklist
*Put an X by all that apply. You can fill this out after submitting the PR. If you have any questions, don't hesitate to ask! We want to help. These are a guide for you to know what the reviewers will be looking for in your contribution.*

- [x] I am submitting my contribution to the develop branch.
- [x] My contribution generates no new compiler warnings (try with the '-Wall -Wextra -Wno-unused-parameter -Wno-empty-body' compiler flags).
- [x] My contribution is commented and consistent with SU2 style.
- [] I have added a test case that demonstrates my contribution, if necessary.
- [] I have updated appropriate documentation (Tutorials, Docs Page, config_template.cpp) , if necessary.

### CODE REVIEW

File: SU2_CFD/include/iteration/CIterationFactory.hpp [hunk 1]
```diff
@@ -0,0 +1,49 @@
...
+   * \brief Deleted constructor to avoid creating instances of this class
+   */
+  CIterationFactory() = delete;
+
+  /*!
+   * \brief Create a new integration instance based on the current sub solver
+   * \param[in] integrationType  - The integration type
+   * \return                     - Pointer to the allocated integration instance
+   */
+  static CIteration* createIteration(ENUM_MAIN_SOLVER kindSolver, const CConfig *config);
```
Comment (economon) [inline SU2_CFD/include/iteration/CIterationFactory.hpp:47]: Does this being a class buy you anything extra? Might just make it a free function. Also, I'd capitalize Create

Comment (talbring) [reply SU2_CFD/include/iteration/CIterationFactory.hpp:47]: I think it is more readable on the caller side if it is in a class. Of course a namespace would be also possible. I think it doesn't really make a difference here (unless you have a strong preference). Gonna change the comments and capitalize the name.

Comment (pcarruscag) [reply SU2_CFD/include/iteration/CIterationFactory.hpp:47]: Class gives you private methods... which you can also emulate with anonymous namespaces same difference in the end.
@talbring is you capitalize this factory then do so for all others please.

File: SU2_CFD/include/iteration/CAdjFluidIteration.hpp [hunk 2]
```diff
@@ -0,0 +1,86 @@
...
+  /*!
+   * \brief Constructor of the class.
+   * \param[in] config - Definition of the particular problem.
+   */
+  CAdjFluidIteration(const CConfig* config);
+
+  /*!
+   * \brief Destructor of the class.
+   */
+  ~CAdjFluidIteration(void) override;
```
Comment (pcarruscag) [inline SU2_CFD/include/iteration/CAdjFluidIteration.hpp:49]: I would delete the empty destructors

File: SU2_CFD/include/iteration/CFluidIteration.hpp [hunk 3]
```diff
@@ -0,0 +1,154 @@
...
+   */
+  void Solve(COutput* output, CIntegration**** integration, CGeometry**** geometry, CSolver***** solver,
+             CNumerics****** numerics, CConfig** config, CSurfaceMovement** surface_movement,
+             CVolumetricMovement*** grid_movement, CFreeFormDefBox*** FFDBox, unsigned short val_iZone,
+             unsigned short val_iInst) override;
+
+  /*!
+   * \brief Updates the containers for the fluid system.
+   * \param[in] ??? - Description here.
+   */
```
Comment (pcarruscag) [inline SU2_CFD/include/iteration/CFluidIteration.hpp:97]: I wonder if we need documentation strings on the derived classes, especially on this class family this stuff is out of date all the time.

File: SU2_CFD/src/iteration/CIteration.cpp [hunk 4]
```diff
@@ -0,0 +1,260 @@
...
+
+  nInst = config->GetnTimeInstances();
+  nZone = config->GetnZone();
+
+  multizone = config->GetMultizone_Problem();
+  singlezone = !(config->GetMultizone_Problem());
+}
+
+CIteration::~CIteration(void) {}
+
```
Comment (pcarruscag) [inline SU2_CFD/src/iteration/CIteration.cpp:45]: Default the destructor in the hpp "the best kind of documentation is the one the compiler understands".

File: SU2_CFD/src/iteration/CIteration.cpp [hunk 5]
```diff
@@ -0,0 +1,260 @@
...
+CIteration::CIteration(const CConfig* config) {
+  rank = SU2_MPI::GetRank();
+  size = SU2_MPI::GetSize();
+
+  nInst = config->GetnTimeInstances();
+  nZone = config->GetnZone();
+
+  multizone = config->GetMultizone_Problem();
+  singlezone = !(config->GetMultizone_Problem());
```
Comment (pcarruscag) [inline SU2_CFD/src/iteration/CIteration.cpp:41]: I would make these members const and initialize in the init list.

File: SU2_CFD/src/iteration/CIteration.cpp [hunk 6]
```diff
@@ -0,0 +1,260 @@
...
+}
+
+void CIteration::Preprocess(COutput* output, CIntegration**** integration, CGeometry**** geometry, CSolver***** solver,
+                            CNumerics****** numerics, CConfig** config, CSurfaceMovement** surface_movement,
+                            CVolumetricMovement*** grid_movement, CFreeFormDefBox*** FFDBox, unsigned short val_iZone,
+                            unsigned short val_iInst) {}
+void CIteration::Iterate(COutput* output, CIntegration**** integration, CGeometry**** geometry, CSolver***** solver,
+                         CNumerics****** numerics, CConfig** config, CSurfaceMovement** surface_movement,
+                         CVolumetricMovement*** grid_movement, CFreeFormDefBox*** FFDBox, unsigned short val_iZone,
+                         unsigned short val_iInst) {}
```
Comment (pcarruscag) [inline SU2_CFD/src/iteration/CIteration.cpp:228]: move to hpp?

File: SU2_CFD/src/iteration/CFluidIteration.cpp [hunk 7]
```diff
@@ -0,0 +1,555 @@
...
+  }
+
+  delete[] Gust;
+  delete[] GustDer;
+  delete[] NewGridVel;
+}
+
+void CFluidIteration::InitializeVortexDistribution(unsigned long& nVortex, vector<su2double>& x0, vector<su2double>& y0,
+                                                   vector<su2double>& vort_strength, vector<su2double>& r_core) {
+  /*--- Read in Vortex Distribution ---*/
```
Comment (pcarruscag) [inline SU2_CFD/src/iteration/CFluidIteration.cpp:501]: Is this needed? I thought we had some vortex stuff in the MMS @economon ?

Comment (economon) [reply SU2_CFD/src/iteration/CFluidIteration.cpp:501]: If I recall, that routine is related to the old implementation of wind gust features, which pre-dates the verification classes. Those initial conditions could probably be moved there (verification), or perhaps @talbring had something in mind with the modules for that

File: SU2_CFD/include/iteration/CIterationFactory.hpp (hunk 1 above)
Comment (economon) [reply SU2_CFD/include/iteration/CIterationFactory.hpp:47]: Fine with either - you decide in the end

File: SU2_CFD/src/iteration/CFluidIteration.cpp (hunk 7 above)
Comment (pcarruscag) [reply SU2_CFD/src/iteration/CFluidIteration.cpp:501]: You are right, for a minute I thought it had something to do with the Taylor-Green vortex.

File: SU2_CFD/include/iteration/CIteration.hpp [hunk 8]
```diff
@@ -0,0 +1,261 @@
...
+   * \param[in] numerics - Description of the numerical method (the way in which the equations are solved).
+   * \param[in] config - Definition of the particular problem.
+   * \param[in] surface_movement - Surface movement classes of the problem.
+   * \param[in] grid_movement - Volume grid movement classes of the problem.
+   * \param[in] FFDBox - FFD FFDBoxes of the problem.
+   */
+  virtual void Predictor(COutput* output, CIntegration**** integration, CGeometry**** geometry, CSolver***** solver,
+                         CNumerics****** numerics, CConfig** config, CSurfaceMovement** surface_movement,
+                         CVolumetricMovement*** grid_movement, CFreeFormDefBox*** FFDBox, unsigned short val_iZone,
+                         unsigned short val_iInst){};
```
Comment (pcarruscag) [inline SU2_CFD/include/iteration/CIteration.hpp:187]: Stray ; on a few of these.

**Pull request merged**